aspnet/migration/http-modules.rst

```diff
@@ -1,8 +1,576 @@
+
+**HttpContext.Request.RequestContext.RouteData** 
+
+RouteData is not available in middleware in RC1.
+
+
+HttpContext.Response
+^^^^^^^^^^^^^^^^^^^^
+
+**HttpContext.Response.Status and HttpContext.Response.StatusDescription** translate to:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Middleware/HttpContextDemoMiddleware.cs
+        :start-after: //>30
+        :end-before: //<30
+        :language: c#
+        :dedent: 12
+
+**HttpContext.Response.ContentEncoding and HttpContext.Response.ContentType** translate to:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Middleware/HttpContextDemoMiddleware.cs
+        :start-after: //>31
+        :end-before: //<31
+        :language: c#
+        :dedent: 12
+
+**HttpContext.Response.ContentType** on its own also translates to:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Middleware/HttpContextDemoMiddleware.cs
+        :start-after: //>32
+        :end-before: //<32
+        :language: c#
+        :dedent: 12
+
+**HttpContext.Response.Output** translates to:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Middleware/HttpContextDemoMiddleware.cs
+        :start-after: //>33
+        :end-before: //<33
+        :language: c#
+        :dedent: 12
+
+Serving up a file is discussed `here <../fundamentals/request-features.html#middleware-and-request-features>`_.
+
+**HttpContext.Response.Headers**
+
+Sending response headers is complicated by the fact that if you set them after anything has been written to the
+response body, they will not be sent.
+
+The solution is to set a callback method that will be called right before writing to the response starts. This is best done at the
+start of the ``Invoke`` method in your middleware. It is this callback method that sets your response headers.
+
+This code sets a callback method called ``SetHeaders``:
+
+    .. code-block:: c#
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            // ...
+            httpContext.Response.OnStarting(SetHeaders, state: httpContext);
+
+Your ``SetHeaders`` callback method would look like this:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Middleware/HttpContextDemoMiddleware.cs
+        :start-after: //>41
+        :end-before: //<41
+        :language: c#
+        :dedent: 8
+
+
+**HttpContext.Response.Cookies**
+
+Cookies travel to the browser in a *Set-Cookie* response header. As a result, sending cookies requires the same callback as used for sending response headers:
+
+    .. code-block:: c#
+
+            public async Task Invoke(HttpContext httpContext)
+            {
+                // ...
+                httpContext.Response.OnStarting(SetCookies, state: httpContext);
+                httpContext.Response.OnStarting(SetHeaders, state: httpContext);
+
+Your ``SetCookies`` callback method would look like this:
```

Comment:
@mperdeck  Wow this is really an awesome doc, very useful and well written. Great work!